boa/src/syntax/lexer/identifier.rs

```diff
@@ -0,0 +1,71 @@
+use super::{Cursor, Error, Tokenizer};
+use crate::syntax::ast::{Position, Span};
+use crate::syntax::lexer::token::Numeric;
+use crate::syntax::lexer::{Token, TokenKind};
+use std::io::Read;
+use std::str::FromStr;
+
+/// Identifier or keyword lexing.
+///
+/// This currently includes boolean/NaN lexing.
+///
+/// More information:
+///  - [ECMAScript reference][spec]
+///  - [MDN documentation][mdn]
+///
+/// [spec]:
+/// [mdn]:
+#[derive(Debug, Clone, Copy)]
+pub(super) struct Identifier {
+    init: char,
+}
+
+impl Identifier {
+    /// Creates a new identifier/keyword lexer.
+    pub(super) fn new(init: char) -> Self {
+        Self { init }
+    }
+}
+
+impl<R> Tokenizer<R> for Identifier {
+    fn lex(&mut self, cursor: &mut Cursor<R>, start_pos: Position) -> Result<Token, Error>
+    where
+        R: Read,
+    {
+        let mut buf = self.init.to_string();
+
+        loop {
+            match cursor.peek() {
+                None => {
+                    break;
+                }
+                Some(Ok(c)) => {
+                    if c.is_alphabetic() || c.is_digit(10) || *c == '_' {
+                        let ch = cursor.next().unwrap()?;
+                        buf.push(ch);
+                    } else {
+                        break;
+                    }
+                }
+                Some(Err(_e)) => {
+                    // TODO handle error.
+                }
+            }
+        }
+        let tk = match buf.as_str() {
+            "true" => TokenKind::BooleanLiteral(true),
+            "false" => TokenKind::BooleanLiteral(false),
+            "null" => TokenKind::NullLiteral,
+            "NaN" => TokenKind::NumericLiteral(Numeric::Rational(f64::NAN)),
```

Comment:
This has been changed in production `NaN` is not a number, but an identifier in the global object.

Reply:
Will need to fix this during the merge

Reply:
Yeah fixed